src/test/java/cat/udl/eps/softarch/demo/steps/StepRegisterAdoption.java

```diff
@@ -0,0 +1,119 @@
+package cat.udl.eps.softarch.demo.steps;
+import cat.udl.eps.softarch.demo.domain.Adoptions;
+import cat.udl.eps.softarch.demo.domain.Cat;
+import cat.udl.eps.softarch.demo.domain.Pet;
+import cat.udl.eps.softarch.demo.domain.User;
+import cat.udl.eps.softarch.demo.repository.AdoptionsRepository;
+import cat.udl.eps.softarch.demo.repository.PetRepository;
+import cat.udl.eps.softarch.demo.repository.UserRepository;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Given;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.Optional;
+
+import static java.lang.Long.parseLong;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.*;
+import static org.mockito.internal.matchers.text.ValuePrinter.print;
+
+public class StepRegisterAdoption {
+
+    @Autowired
+    private StepDefs stepDefs;
+
+    @Autowired
+    private PetRepository petRepository;
+
+    @Autowired
+    private AdoptionsRepository adoptionsRepository;
+
+    @Autowired
+    private UserRepository userRepository;
+
+
+
+    @Given("^The Pet with id (\\d+) does not exists")
+    public void nonExistingPet (int id_str) {
+        Long id = (long) id_str;
+        Pet pet = petRepository.findPetById(id);
+        assertNull(pet);
+    }
+
+    @Given("^The Pet with id (\\d+) is not adopted")
+    public void isNotAdopted (int id_int) {
+        Long id = (long) id_int;
+        Pet pet = petRepository.findPetById(id);
+        assertTrue(pet != null && !pet.isAdopted());
+    }
+
+    @Given("^The Pet with id (\\d+) is adopted by user \"([^\"]*)\"")
+    public void isAdopted (int id_int, String username) {
+        Long id = (long) id_int;
+        Pet pet = petRepository.findPetById(id);
+        Adoptions adoptions = new Adoptions();
+        userRepository.findById(username).ifPresent(user -> adoptions.setUser(user));
+        adoptions.setPet(pet);
+        pet.setAdopted(true);
+        petRepository.save(pet);
+        assertTrue(pet.isAdopted());
+    }
+
+
+    @When("^The user with username \"([^\"]*)\" adopts the Pet with id (\\d+)")
```

Comment:
The When step should be testing the API using the mockMvc object, otherwise it is not tested at all. Consequently, you should be here doing a POST to /adoptions to create an Adoption for the pet (identified by its name, not the id). Then, in the application code (not in the tests), use the Adoption handler and implement the HandleBeforeLinkSave when an adoption is linked to a pet, check that that pet is not yet adopted and then update its status to adopted